daemon/network.go

```diff
@@ -893,6 +893,13 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
 
 	createOptions = append(createOptions, libnetwork.EndpointOptionGeneric(genericOptions))
 
+	// IPv6 may be enabled on the network, but disabled in the container.
+	if n.IPv6Enabled() {
+		if sbIPv6, ok := sb.IPv6Enabled(); ok && !sbIPv6 {
+			createOptions = append(createOptions, libnetwork.CreateOptionDisableIPv6())
+		}
+	}
```

Comment:
The endpoint could determine for itself whether IPv6 is enabled in the sandbox when the Endpoint is Joined to the Sandbox. It'd make sense to have an endpoint option persisted in the store for _the user_ to disable IPv6 allocation for an attachment, but it feels weird to use an endpoint option to convey a fact about the sandbox it will be joined to. The endpoint option might get out of sync with ground truth.

How feasible would it be to defer IPAM allocation until the join operation?

Reply:
I think it's feasible - and I agree it'd be better, but it would be a behavioural change for libnetwork. At the moment, our daemon code expects to be able to stash address info in NetworkSettings after creating the endpoint, before the join. That can be changed but, do we need to worry about other users of libnetwork doing something similar?

Alternatively, we could pass the `libnetwork.Sandbox` to the `CreateEndpoint` as an `EndpointOption` - not to be stored in the `Endpoint`, just so it can see what it'll be joined to? (If it gets no sandbox option, or the sandbox doesn't have an osSbox that can be checked for the `::1`, the Endpoint would just follow its current rules for allocating IPv6 addresses - and libnetwork's behaviour wouldn't change.)

Reply:
> Alternatively, we could pass the libnetwork.Sandbox to the CreateEndpoint as an EndpointOption ...

Expanding on that idea, I added `CreateOptionJoinSandbox` ... nothing creates an Endpoint without adding it to a Sandbox. Optionally combining the two steps means the CreateEndpoint has access to the Sandbox, so it can inspect it.

Reply:
> > At the moment, our daemon code expects to be able to stash address info in NetworkSettings after creating the endpoint, before the join. That can be changed
> 
> Stash info for _which_ addresses? IPAM-allocated addresses are allocated without the daemon in the loop, IIUC. And I'm not sure why changing when IPAM addresses are allocated _inside libnetwork_ has any impact on daemon code.

The existing daemon code does ...
  - CreateEndpoint()
    - _libnetwork allocates addresses_
  - Copy the addresses libnetwork just allocated into NetworkSettings.
  - Join()

If libnetwork delayed addresses allocation until the Join(), NetworkSettings would have to be updated after the Join(). That's ok, but it is a change needed in the daemon because of an hidden change in the behaviour of libnetwork's API.

> > Optionally combining the two steps means the CreateEndpoint has access to the Sandbox, so it can inspect it.
> 
> CreateEndpoint would have access to _that instance_ of the Sandbox to inspect _the OS state at that point in time._ I'm worried about the restore story, where a container running when the daemon is stopped is restarted when the daemon is next started. Consider the case where the system is rebooted in between and `disable_ipv6=true` is added to or removed from the kernel command line. When the daemon is started, the existing Endpoint and Sandbox are rehydrated from libnetwork's datastore. If IPv6 support in the sandbox is only checked when the endpoint is created, the container would not get an IPv6 address if IPv6 support was enabled in the kernel between reboots, or an IPv6 address would be allocated that does not get used if IPv6 support was disabled. Actually, deferring IPAM allocation to join would not be sufficient to solve this restore case as sandbox restore follows a different code-path from endpoint join. IPAM allocation would also have to be redone on restore. (Redoing allocation on restore would also be required to e.g. properly implement a DHCP-client IPAM allocator.)

Yes, 'restore' restores a container with the config it had before. Re-creating a container and disconnect/connect are the ways to change its networking config.

Allocating addresses during Join(), and then not actually just Join() because that wouldn't help with the restore case anyway, would be a further complicated and delicate change - a lot of code paths are involved. So, while dealing with a requirement to change the host's networking stack without re-creating containers might be worthwhile - I think it implies significantly more risk and work, and doesn't need to be part of this change.

This PR cleans up construction of Sandbox and Endpoints by separating them cleanly, it doesn't move any responsibility for configuration of legacy links between the daemon and libnetwork, and it makes it possible to attach an IPv4 container to an IPv6 network without it getting IPv6 addresses.

Reply:
This also paves the way for not-allocating IPv4 addresses for an IPv6 only network, which is possibly more important.  There's quite a bit more to sort out for that, but this is a first step.

Reply:
I think it's safe to say there are no other users of libnetwork for container networking. Buildkit only uses the resolvconf package and any third-party users are doing so at their own risk, IMO.